packages/@aws-cdk/aws-iam/lib/optimize-statements.ts

```diff
@@ -0,0 +1,286 @@
+   * Verifies if a list of entities is entirely covered by another.
+   * @param left  a list of de-duplicable entities
+   * @param right a list of de-duplicable entities.
+   * @returns ``true`` if all entries from ``left`` are superceded by at least one entry from ``right``.
+   */
+  private _allSuperceded<T extends Deduplicable>(left: T[], right: T[]): boolean {
+    if (left.length === 0) { return right.length === 0; }
+    return left.find(item => right.find(e => e.supercedes(item)) == null) == null;
+  }
+}
+
+/**
+ * A glob expression, which can be either a string with ``*`` wildcards or a CloudFormation intrinsic.
+ */
+class Glob implements Deduplicable {
+  private regexp?: RegExp;
+
+  /**
+   * @param expression the expression from the policy statement.
+   */
+  constructor(public readonly expression: unknown) {
+    this.regexp = _toRegExp(this.expression);
+  }
+
+  public supercedes(other: Glob): boolean {
+    if (!this.regexp || typeof other.expression !== 'string') {
+      return _deepEqual(this.expression, other.expression);
+    }
+    return this.regexp.test(other.expression)
+      // Use the shortest expression (** matches what * does, and * is better)
+      && (this.expression as string).length <= other.expression.length;
+  }
+}
+
+/**
+ * Entities that can be de-duplicated
+ */
+interface Deduplicable {
+  /**
+   * Checks whether an entity can safely replace another (ie: it is a superset).
+   * @param other the other entity to check.
+   * @returns ``true`` if this entry supercedes (can safely replace) ``other``.
+   */
+  supercedes(other: this): boolean;
+}
+
+/**
+ * Entities that can be merged.
+ */
+interface Mergeable {
+  /**
+   * Checks whether an entity is compatible for merging in this one.
+   * @param other the other entity to check.
+   * @returns ``true`` if ``other`` can safely be merged into this entity.
+   */
+  mergeable(other: this): boolean;
+  /**
+   * Merges ``other`` in this entity. It is the caller's responsibility to have
+   * verified the safety of doing so by having called ``mergeable`` before.
+   * @param other the entity to merge in.
+   */
+  merge(other: this): void;
+}
+function _isMergeable<T>(x: T): x is T & Mergeable {
+  return x != null
+    && typeof (x as any).mergeable === 'function'
+    && typeof (x as any).merge === 'function';
+}
+
+/**
+ * Removes duplicate entities from a list.
+ * @param array the list to be de-duplciated.
+ * @returns a new list containing the minimum set of entities.
+ */
+function _deduplicate<T extends Deduplicable>(array: T[]): T[] {
+  const result = new Array<T>();
+  for (let i = 0 ; i < array.length ; i++) {
+    const element = array[i];
+    let redundant = false;
+    for (let j = i + 1 ; j < array.length ; j++) {
+      const candidate = array[j];
+      if (candidate.supercedes(element)) {
+        redundant = true;
+        break;
+      } else if (_isMergeable(candidate) && _isMergeable(element) && candidate.mergeable(element)) {
```

Comment:
Why make it hard on yourself by having a distinction between `Deduplicable` and `Mergeable`? Either two objects can be merged into one without losing meaning, or they can't.

Rewritten without the boolean (simpler, imo) and without the distinction between the two classes:

```
const result = new Array<T>();

while (array.length > 0) {
    result.push(array.splice(0, 1)[0]);
    const last = result[result.length - 1];

    let i = 0;
    while (i < array.length) {
        if (last.mergeable(array[i])) {
            last.merge(array.splice(i, 1)[0]);
        } else {
            i++;
        }
    }
}

return result;
```

Reply:
Thanks :)